gilbreth_perception/src/recognition_node.cpp

```diff
@@ -0,0 +1,295 @@
+#include <ros/ros.h>
+#include <sensor_msgs/PointCloud2.h>
+#include <pcl_conversions/pcl_conversions.h>
+#include <pcl/point_types.h>
+#include <pcl/segmentation/sac_segmentation.h>
+#include <pcl/segmentation/extract_clusters.h>
+#include <pcl/io/pcd_io.h>
+#include <pcl/point_cloud.h>
+#include <pcl/correspondence.h>
+#include <pcl/features/normal_3d_omp.h>
+#include <pcl/features/shot_omp.h>
+#include <pcl/features/board.h>
+#include <pcl/features/normal_3d.h>
+#include <pcl/features/fpfh.h>
+#include <pcl/keypoints/uniform_sampling.h>
+#include <pcl/visualization/pcl_visualizer.h>
+#include <pcl/kdtree/kdtree_flann.h>
+#include <pcl/kdtree/impl/kdtree_flann.hpp>
+#include <pcl/common/transforms.h>
+#include <pcl/console/parse.h>
+#include <pcl/sample_consensus/method_types.h>
+#include <pcl/sample_consensus/model_types.h>
+#include <pcl/ModelCoefficients.h>
+#include <pcl/filters/extract_indices.h>
+#include <pcl/filters/passthrough.h>
+#include <pcl/filters/voxel_grid.h>
+#include <pcl/registration/ia_ransac.h>
+#include <tf/transform_listener.h>
+#include <geometry_msgs/PointStamped.h>
+#include <iostream>
+#include <cstdio>
+#include <fstream>
+#include <gilbreth_perception/ObjectDetection.h>
+
+typedef pcl::PointXYZ PointType;
+typedef pcl::Normal NormalType;
+
+//Algorithm params
+float descr_rad_ (0.02f);
+float down_sample_(0.01);
+float min_sample_distance_ (0.025f);
+float max_correspondence_distance_ (0.01f*0.01f);
+int nr_iterations_ (20);
+
+
+pcl::visualization::PCLVisualizer viewer ("recognition_result");
+ros::Publisher pub_tf, pub;
+
+
+struct Result
+{
+	std::string item_name;
+	int item_id;
+	float fitness_score;
+	pcl::PointXYZ pick_point;
+	Eigen::Matrix4f final_transformation;
+	EIGEN_MAKE_ALIGNED_OPERATOR_NEW
```

Comment:
It is from one PCL example, their explain is:
Because we are including an Eigen::Matrix4f in this struct, we need to include the EIGEN_MAKE_ALIGNED_OPERATOR_NEW macro, which will overload the struct’s “operator new” so that it will generate 16-bytes-aligned pointers.
Link of the PCL example:http://pointclouds.org/documentation/tutorials/template_alignment.php#template-alignment